src/decode/lzma.rs

```diff
@@ -264,76 +315,206 @@ where
 
-                len = self.rep_len_decoder.decode(rangecoder, pos_state)?;
+            if update {
                 // update state (rep)
                 self.state = if self.state < 7 { 8 } else { 11 };
-            // New distance
-            } else {
+            }
+        // New distance
+        } else {
+            if update {
                 // Update LRU
                 self.rep[3] = self.rep[2];
                 self.rep[2] = self.rep[1];
                 self.rep[1] = self.rep[0];
-                len = self.len_decoder.decode(rangecoder, pos_state)?;
+            }
 
+            len = self.len_decoder.decode(rangecoder, pos_state, update)?;
+
+            if update {
                 // update state (match)
                 self.state = if self.state < 7 { 7 } else { 10 };
-                self.rep[0] = self.decode_distance(rangecoder, len)?;
+            }
 
+            let rep_0 = self.decode_distance(rangecoder, len, update)?;
+
+            if update {
+                self.rep[0] = rep_0;
                 if self.rep[0] == 0xFFFF_FFFF {
                     if rangecoder.is_finished_ok()? {
-                        break;
+                        return Ok(ProcessingStatus::Finished);
                     }
                     return Err(error::Error::LZMAError(String::from(
                         "Found end-of-stream marker but more bytes are available",
                     )));
                 }
             }
+        }
 
+        if update {
             len += 2;
 
             let dist = self.rep[0] + 1;
             self.output.append_lz(len, dist)?;
         }
 
+        Ok(ProcessingStatus::Continue)
+    }
+
+    fn process_next<'a, R: io::BufRead>(
+        &mut self,
+        rangecoder: &mut rangecoder::RangeDecoder<'a, R>,
+    ) -> error::Result<ProcessingStatus> {
+        self.process_next_inner(rangecoder, true)
+    }
+
+    /// Try to process the next iteration of the loop.
+    ///
+    /// This will check to see if there is enough data to consume and advance the
+    /// decompressor. Needed in streaming mode to avoid corrupting the state while
+    /// processing incomplete chunks of data.
+    fn try_process_next(&mut self, buf: &[u8], range: u32, code: u32) -> error::Result<()> {
+        let mut temp = std::io::Cursor::new(buf);
+        let mut rangecoder = rangecoder::RangeDecoder::from_parts(&mut temp, range, code);
+        let _ = self.process_next_inner(&mut rangecoder, false)?;
+        Ok(())
+    }
+
+    /// Utility function to read data into the partial input buffer.
+    fn read_partial_input_buf<'a, R: io::BufRead>(
+        &mut self,
+        rangecoder: &mut rangecoder::RangeDecoder<'a, R>,
+    ) -> error::Result<()> {
+        // Fill as much of the tmp buffer as possible
+        let start = self.partial_input_buf.position() as usize;
+        let bytes_read =
+            rangecoder.read_into(&mut self.partial_input_buf.get_mut()[start..])? as u64;
```

Comment:
CI was failing because Rust 1.32 doesn't have `std::convert::TryFrom`. I think the cast to u64 is safe here because of the invariant `bytes_read <= MAX_REQUIRED_INPUT`.